### PR TITLE
Don't include completions for current and later parameters

### LINKED ISSUE #1858: Only prior declared parameters should appear in completion lists for initializers
``` TypeScript
function f(x = /**/, y) {
}
```

At `/**/`, `x` and `y` should not appear in the completion list, but they currently do.

### LINKED ISSUE #4046: Should not show completion when properties of type literals/interfaces span different lines
completionintypeliterals

### PR DESCRIPTION
**Pull request by zardoy**
Tried to find oldest completion issue :)

Fixes #1858

### CODE REVIEW

File: src/services/completions.ts
Comment (DanielRosenwasser) [inline src/services/completions.ts:2142]: I wonder if you could just get await with using the node positions.

```suggestion
                const declarationPos = symbol.valueDeclaration.pos;
                if (parameterDeclaration.pos <= declarationPos && declarationPos < parameters.end) {
                    return false;
                }
```

File: tests/cases/fourslash/noIncorrectParamaterCompletions.ts
Comment (DanielRosenwasser) [inline tests/cases/fourslash/noIncorrectParamaterCompletions.ts:5]: These aren't valid declarations of methods or functions.

Comment (DanielRosenwasser) [reply tests/cases/fourslash/noIncorrectParamaterCompletions.ts:5]: Also, it would be good to test out that within `b`, you still do get completions for `a`.

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]
The test file should be renamed:

```diff
tests/cases/fourslash/noIncorrectParamaterCompletions.ts
tests/cases/fourslash/noCompletionsForCurrentOrLaterParameters.ts
```

File: src/services/completions.ts
Comment (zardoy) [reply src/services/completions.ts:2053]: So your decision is to unify `variableDeclaration` and `parameterDeclaration`?

eg

```ts
const variableOrParameterDeclaration = getVariableOrParameterDeclaration(location, contextToken)
```

But what about this case then?:

```ts
const foo = (a = foo/*no foo suggestion*/) => {
    console.log(foo)
}
```

Comment (zardoy) [reply src/services/completions.ts:2053]: BTW its fully valid location for `foo` suggestion, so maybe you're right

File: tests/cases/fourslash/noIncorrectParamaterCompletions.ts
Comment (zardoy) [reply tests/cases/fourslash/noIncorrectParamaterCompletions.ts]: Oops was too focused on ide testing, now fixed

File: src/services/completions.ts
Comment (zardoy) [reply src/services/completions.ts]: Seems to be good idea! but I still wonder why do we need `declarationPos < parameters.end` check?

**Comment by zardoy** [comment]
Didn't expect so many failing tests 😅 , will look again at them in a few days

File: src/services/completions.ts
Comment (zardoy) [reply src/services/completions.ts]: I switched to `getVariableOrParameterDeclaration(contextToken: Node | undefined)`. I'm using only `contextToken` to fit logic in single `findAncestor`. I don't really the idea of dropping `location` as I feel it can introduce a lot of breakage, though I'll look at tests and try to fix them

Comment (zardoy) [reply src/services/completions.ts]: > BTW its fully valid location for foo suggestion, so maybe you're right

FYI https://github.com/microsoft/TypeScript/blob/0791bb0f8405058877c9930d63457f32522ae27c/tests/cases/fourslash/completionListWithoutVariableinitializer.ts#L50

**Comment by zardoy** [comment]
It seems using contextToken in new implementation improved quality in some test cases, I'm happy with result

**Comment by zardoy** [comment]
Also I have a quick question. Speaking of `getVariableOrParameterDeclaration` I wonder wether the logic can be extracted for codefixes in future. Just for motivation consider this issue:

```ts
let fileName

const foo = (displayName = fileNam/*<-*/, fileName2) => {
    // ERROR: Cannot find name 'fileNam'. Did you mean 'fileName2'?

    // super annoying sometimes, would be cool if codefixes were smarter
}
```

**Comment by zardoy** [comment]
@DanielRosenwasser I'm sorry to bother you, but I wanted to do the same for type parameters. Would it be better to do it in this PR or create another one after its get merged? Its just a matter of adding an extra check 😉.

File: src/services/completions.ts
Comment (DanielRosenwasser) [inline src/services/completions.ts:2133]: I would just write `variableOrParameterDeclaration && isParameter(variableOrParameterDeclaration) && symbolDeclaration && isParameter(symbolDeclaration)` just so you can avoid the type assertions below.

Comment (zardoy) [reply src/services/completions.ts]: Not a problem!

**Comment by jakebailey** [comment]
> I'm sorry to bother you, but I wanted to do the same for type parameters. Would it be better to do it in this PR or create another one after its get merged? Its just a matter of adding an extra check 😉.

How simple are we talking? If it's not too bad, I don't think it's a big deal to include it here.

File: tests/cases/fourslash/completionListIsGlobalCompletion.ts
Comment (zardoy) [inline tests/cases/fourslash/completionListIsGlobalCompletion.ts]: All done, but I'm kind of clueless why it wasnt failing on CI 2 commits ago

**Comment by zardoy** [comment]
> How simple are we talking? If it's not too bad, I don't think it's a big deal to include it here.

Okay, I think it's done

File: src/services/completions.ts
Comment (sandersn) [inline src/services/completions.ts:5125]: the last check isn't needed because there's only one child of an arrow function that can have kind EqualsGreaterThanToken.

File: src/services/completions.ts
Comment (sandersn) [inline src/services/completions.ts:5118]: 1. The `"quit"` check should be first.
2. When the parameter checks are moved down with the variable check, the line becomes long enough that the original two-line temp+cast becomes easier to read again.

```suggestion
    const declaration = findAncestor(contextToken, node =>
        isFunctionBlock(node) || isArrowFunctionBody(node) || isBindingPattern(node)
            ? "quit"
            :  isVariableDeclaration(node) || ((isParameter(node) || isTypeParameterDeclaration(node)) && !isIndexSignatureDeclaration(node.parent)));
    return declaration as ParameterDeclaration | TypeParameterDeclaration | VariableDeclaration | undefined;
```

File: tests/cases/fourslash/completionListInUnclosedFunction09.ts
Comment (sandersn) [inline tests/cases/fourslash/completionListInUnclosedFunction09.ts:8]: Why didn't this work before? There was already code for it and you didn't change it very much.

File: src/services/completions.ts
Comment (sandersn) [inline src/services/completions.ts:2146]: Use if-then/if-then/else formatting for conditionals:

```suggestion
                const parameters = isParameter(variableOrParameterDeclaration) ? variableOrParameterDeclaration.parent.parameters :
                    isInferTypeNode(variableOrParameterDeclaration.parent) ? undefined :
                    variableOrParameterDeclaration.parent.typeParameters;
```

File: src/services/completions.ts
Comment (sandersn) [inline src/services/completions.ts:2147]: I don't understand any of these predicates.
- why not check `symbolDeclaration === variableOrParameterDeclaration`?
- why check `parameters`?
- when can symbolDeclaration.pos be larger than parameters.end?

**Review by sandersn — CHANGES_REQUESTED** [review]
Very nice and small change. I had some questions and suggestions.

File: src/services/completions.ts
Comment (sandersn) [reply src/services/completions.ts:2147]: reading the previous reviews, it seems like this code was a suggestion from @DanielRosenwasser

Comment (sandersn) [reply src/services/completions.ts:2145]: @DanielRosenwasser can you explain this check?

Comment (DanielRosenwasser) [reply src/services/completions.ts:2145]: Yeah - it seemed unnecessary to run through all the parameters to find the parameter index and compare with that. Just checking the textual position seemed like enough. So we just check if the symbol was declared *somewhere after* the current parameter, but still within the parameter list.

Comment (DanielRosenwasser) [reply src/services/completions.ts:2145]: Also, I'm pretty sure you can't really do a naive walk of the parameter list because of destructuring.

Comment (DanielRosenwasser) [reply src/services/completions.ts:2145]: > when can symbolDeclaration.pos be larger than parameters.end

Technically we could remove that check - you probably don't want to provide completions for *any* declaration even if it's within the parameter list.

Comment (DanielRosenwasser) [reply src/services/completions.ts:2145]: Well...wait, that's not true.

```ts
function f(x = /**/) {
}

var y;
```

You still want completions for `y` there. So that's what I was thinking.

File: tests/cases/fourslash/completionListInUnclosedFunction09.ts
Comment (zardoy) [reply tests/cases/fourslash/completionListInUnclosedFunction09.ts:8]: https://github.com/microsoft/TypeScript/pull/52690#issuecomment-1424655384 I guess thats because I'm now using `contextToken` instead of `location`, I didn't really dig into specific test cases, but isnt it an improvement? Also see https://github.com/microsoft/TypeScript/pull/52690#discussion_r1101825361

File: src/services/completions.ts
Comment (zardoy) [reply src/services/completions.ts:2145]: > You still want completions for `y` there. So that's what I was trying to model.

But we have the following check: `isParameter(variableOrParameterDeclaration) && isParameter(symbolDeclaration)` line: https://github.com/microsoft/TypeScript/pull/52690/files#diff-4237fd08a5f920543e184b7748bc8ae000d9b1c031810ba375b0053362ec2b6aR2139, since `y` is not parameter it doesn't get included anyway

Comment (sandersn) [reply src/services/completions.ts:2145]: If what's there now works, I think it's fine. It's pretty understandable and more general than strict equality of declarations.

**Review by sandersn — APPROVED** [review]

**Comment by zardoy** [comment]
I see it's in waiting on author. Is anything else required here from my side?

**Comment by sandersn** [comment]
I was waiting on a review from @DanielRosenwasser since he previously requested changes.

**Review by DanielRosenwasser — APPROVED** [review]

**Pull request merged**
